### ISSUE TITLE
Newly joined nodes don't receive all routes when using RR mode

### ISSUE THREAD

**Issue by Feder1co5oave** [cloudnativelabs/kube-router#676]
On an already existing kubernetes cluster with kube-router successfully running in route reflector mode (`rr.server` annotation on 3 master nodes, `rr.client` annotation on all the workers), my workflow for joining new worker nodes is like this (the cluster was created with kubeadm):

for each new node:
- run kubeadm join on the node
- kubectl annotate node <new-node> kube-router.io/rr.client=23
- restart kube-router on the new node to put them in RR mode

Turns out the nodes previously joined to the cluster will receive routes for all the new nodes, but the new nodes will only receive routes for the "old" nodes. Restarting kube-router on the route reflector nodes solved this issue.

My troubleshooting suggested that whenever a new node joins, the rr server kube-router daemons will peer with that node right away, even before the `rr.client` annotation gets added to it. Thus, the new peer is assumed to be in full mesh mode even if in fact it will be in RR mode and will only peer with RR servers. BGP route reflector allows for BGP daemons in the same AS to peer in either full-mesh or RR mode. Export policies are such that RR servers will reflect advertisements:
- from RR peers to RR peers
- from full-mesh peers to RR peers
- from RR peers to full-mesh peers
- but NOT from full-mesh peers to other full-mesh peers (they're supposed to peer with each other in full-mesh, duh)

So the new nodes, assumed to be forming full-mesh, don't get advertisements about other new nodes.
Restarting RR servers forces them to reload the node list and annotations, and correctly peer with the new nodes in RR mode.

To fix this, kube-router should watch for annotation changes on nodes, and update its internal information about which nodes are forming full-mesh, and which are joined to a RR cluster.
While waiting for a fix to be implemented, I suggest the workaround to restart RR servers be documented for newcomers!

**Comment by aauren** [comment]
Closed via #677

**Comment by ofen** [comment]
Is it possible to make it automatically (without manual POD restart)?

**Comment by murali-reddy** [comment]
@ofen reopening as still below make sense to be fixed.

> kube-router should watch for annotation changes on nodes, and update its internal information about which nodes are forming full-mesh, and which are joined to a RR cluster.
While waiting for a fix to be implemented, I suggest the workaround to restart RR servers be documented for newcomers!

**Comment by zhaixigui** [comment]
Why can't it automatically recognize the newly added node? And to propagate routing。Why have to do such a dangerous operation to kill RR?

**Comment by zhaixigui** [comment]
Do we have any choice but to kill the RR?

**Comment by Feder1co5oave** [comment]
@zhaixigui the new nodes are recognized, but they are assumed to be in full mesh mode, and there's no clean way to do the switch to RR mode other than restart kube-router on both RR server and client to reload the nodes' information.

Killing RR servers should not bring any disruption as long as you enabled soft-restart. I've done it several times without any repercussions.

**Comment by zhaixigui** [comment]
@Feder1co5oave   a bad idea， when kube-router watched a ”kube-router.io/rr.client=42“ annotation from
most recently added node.  it first calls method DeleteNeighbor to delete this node, and then calls method AddNeighbor to add this node as RR client, and RR will automatically propagate this route ？

**Comment by zhaixigui** [comment]
> @zhaixigui the new nodes are recognized, but they are assumed to be in full mesh mode, and there's no clean way to do the switch to RR mode other than restart kube-router on both RR server and client to reload the nodes' information.
> 
> Killing RR servers should not bring any disruption as long as you enabled soft-restart. I've done it several times without any repercussions.

soft-restart is GracefulRestart ?

**Comment by zhaixigui** [comment]
> @Feder1co5oave a bad idea？ when the RR server watched a ”kube-router.io/rr.client=42“ annotation from
> most recently joined node. it first calls method DeleteNeighbor to delete this node, and then calls method AddNeighbor to add this node as RR client, and RR will automatically propagate this route ？

@Feder1co5oave Feder， is this a bad idea?

**Comment by Feder1co5oave** [comment]
@zhaixigui not at all, in fact it is exactly the solution I proposed in my first post. But I'm not a gopher and this would involve a bit of modifications to the control flow, i.e. watching nodes. I believe we'll have to wait for a volunteer to write a PR.

**Comment by sebltm** [comment]
I think it's as easy as doing something like this: https://github.com/cloudnativelabs/kube-router/pull/1723
I've tested it in my local environment, seems to work as expected

**Comment by aauren** [comment]
Confirmed this is broader than just annotations. `NodeEventHandler.UpdateFunc` in `bgp_peers.go` is an explicit no-op ("we are only interested in node add/delete, so skip update"), so it's not just `rr.client`/`rr.server` that gets missed on UPDATE, it's address changes and pod CIDR changes too.

Worse, `AddFunc` also drops events: it calls `utils.NewRemoteKRNode(node)` and returns early on error, but that's guaranteed to fail on any cloud with an out-of-tree CCM, because the kubelet leaves `node.Status.Addresses` empty until the CCM initializes the node. On a six node kOps cluster on AWS this produced a five minute outage on every bring-up: the control-plane node ended up with zero BGP peers, pods couldn't reach the API server, and only the 5-minute periodic `syncInternalPeers()` eventually recovered it.

I've written up a fix (predicate-driven `UpdateFunc` covering addresses/pod CIDR/`kube-router.io/` annotations, plus dropping the early return in `AddFunc`) and I'm working through it now.
